RecoTracker/TkSeedingLayers/interface/SeedingHitSet.h

```diff
@@ -11,40 +19,60 @@ class SeedingHitSet {
 
   static ConstRecHitPointer nullPtr() { return nullptr; }
 
-  SeedingHitSet() { theRecHits[0] = theRecHits[1] = theRecHits[2] = theRecHits[3] = nullptr; }
+  SeedingHitSet() { initContainer(); }
 
-  SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two)
-  // : theRecHits{{one,two,ConstRecHitPointer()}}
-  {
+  SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two) {
+    initContainer();
     theRecHits[0] = one;
     theRecHits[1] = two;
-    theRecHits[2] = theRecHits[3] = nullptr;
+    setSize();
   }
-  SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two, ConstRecHitPointer three)
-  // : theRecHits{{one,two,three}},
-  {
+  SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two, ConstRecHitPointer three) {
+    initContainer();
     theRecHits[0] = one;
     theRecHits[1] = two;
     theRecHits[2] = three;
-    theRecHits[3] = nullptr;
+    setSize();
   }
 
   SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two, ConstRecHitPointer three, ConstRecHitPointer four) {
+    initContainer();
     theRecHits[0] = one;
     theRecHits[1] = two;
     theRecHits[2] = three;
     theRecHits[3] = four;
+    setSize();
+  }
+
+  SeedingHitSet(std::vector<ConstRecHitPointer> &hits) {
+    if (hits.size() > seedinghits::kMaxSeedSize)
+      edm::LogWarning("SeedingHitSet") << "More than " << seedinghits::kMaxSeedSize
+                                       << " hits give in input. Taking only the first " << seedinghits::kMaxSeedSize;
+
+    initContainer();
+    std::copy(hits.begin(), hits.begin() + std::min(hits.size(), size_t(seedinghits::kMaxSeedSize)), theRecHits);
+    setSize();
   }
 
   ConstRecHitPointer const *data() const { return theRecHits; }
 
-  unsigned int size() const { return theRecHits[3] ? 4 : (theRecHits[2] ? 3 : (theRecHits[1] ? 2 : 0)); }
+  unsigned int size() const { return theSize; }
 
   ConstRecHitPointer get(unsigned int i) const { return theRecHits[i]; }
   ConstRecHitPointer operator[](unsigned int i) const { return theRecHits[i]; }
 
 private:
-  ConstRecHitPointer theRecHits[4];
+  ConstRecHitPointer theRecHits[seedinghits::kMaxSeedSize];
+  unsigned int theSize;
+
+  void setSize() {
+    theSize = 0;
+    while (theRecHits[++theSize])
+      ;
+    theSize = theSize > 1 ? theSize : 0;
+  }
+
+  void initContainer() { std::fill(theRecHits, theRecHits + seedinghits::kMaxSeedSize, SeedingHitSet::nullPtr()); }
```

Comment:
is this method/call actually needed?
The `ConstRecHitPointer` is a `std::shared_ptr<TrackingRecHit const>`. IIUC, shared_ptr come default-initialized (nullptr)

Reply:
uhm, what was the actual size in that case?
The `setSize` code is constructed in a way that if the actual size is `kMaxSeedSize`, it will keep looping out of range. 
Please add a check `theSize < kMaxSeedSize`

Reply:
Should have been `3` since the `SeedingHitSet(ConstRecHitPointer one, ConstRecHitPointer two, ConstRecHitPointer three)` constructor (with three valid hits).